### PR TITLE
Hide testing functions within appropriate CPP macros

### PR DESCRIPTION
**Pull request by Dr15Jones**
Functions which are only used during debugging were being reported
by a clang warning as unused. Moving them into the same CPP macros
as their calling code makes them only available if they will be
called.

### CODE REVIEW

**Comment by davidlange6** [comment]
please test

File: DataFormats/TrackerRecHit2D/src/BaseTrackerRecHit.cc [hunk 1]
```diff
@@ -6,6 +6,7 @@
 
 
 namespace {
+#ifdef EDM_LM_DEBUG
```
Comment (makortel) [inline DataFormats/TrackerRecHit2D/src/BaseTrackerRecHit.cc]: Should this be `EDM_ML_DEBUG`?

Comment (Dr15Jones) [reply DataFormats/TrackerRecHit2D/src/BaseTrackerRecHit.cc]: It is the same as the CPP macro name used later in the file:
https://github.com/cms-sw/cmssw/blob/master/DataFormats/TrackerRecHit2D/src/BaseTrackerRecHit.cc#L23

**Comment by perrotta** [comment]
please test
(AddOn error seems untrelated)

File: DataFormats/TrackerRecHit2D/src/BaseTrackerRecHit.cc (hunk 1 above)
Comment (slava77) [reply DataFormats/TrackerRecHit2D/src/BaseTrackerRecHit.cc]: I think that this and DataFormats/TrackerRecHit2D/interface/BaseTrackerRecHit.h should have a change to EDM_ML_DEBUG. This appears to be a typo.

About wrapping this solely into just ```EDM_ML_DEBUG```, I note the commented out calls to ```throwExceptionUninitialized``` which are perhaps also needed for other non-```EDM_ML_DEBUG``` debugging.
Similar to TrackerSingleRecHit.cc, I'd like to suggest a flag (e.g. ```DO_THROW_UNITITIALIZED```) and put the currently commented out calls in e.g. ```getKfComponents1D``` behind that flag.

**Comment by Dr15Jones** [comment]
please test

File: DataFormats/TrackerRecHit2D/src/BaseTrackerRecHit.cc [hunk 2]
```diff
@@ -1,3 +1,4 @@
+//#define DO_THROW_UNINITIALIZED
```
Comment (slava77) [inline DataFormats/TrackerRecHit2D/src/BaseTrackerRecHit.cc]: shouldn't this be in .h file?
If define is added only here, then in all cases where the compiler decides to not inline the call of ```check``` where it's used, there will be a problem with one definition rule.

**Comment by slava77** [comment]
@Dr15Jones 
your updates will now lead to a different behavior compared to what the affected code had previously.

before
- ```EDM_LM_DEBUG``` was needed to enable non-trivial ```check``` implementation with ```throwExceptionUninitialized``` call
- uncommenting of ```//if (!hasPositionAndError()) throwExceptionUninitialized``` was required to enable the check in additional methods

proposed:
- ```DO_THROW_UNINITIALIZED``` enables all cases with calls to ```throwExceptionUninitialized``` , including the previously uncommented ones.

I would expect that what was previously enabled by ```EDM_LM_DEBUG``` to be enabled by ```EDM_ML_DEBUG``` and the additional (currently commented out) calls would require the extra  ```DO_THROW_UNINITIALIZED``` to be set

**Comment by Dr15Jones** [comment]
@slava77 I'm very much against using EDM_ML_DEBUG to switch the behavior of `check()` as it will too easily cause one definition rules. For example, say one does not checkout DataFormats/TrackerRecHit2D but does have a module that uses it checked out. If EDM_ML_DEBUG is turned on for building that module, the header file will declare a different behavior from the libDataFormatsTrackerRecHit2D.so and we could get strange behaviors.

**Comment by Dr15Jones** [comment]
The handling of previously commented versus macro enabled is now done via two different macros.

**Comment by davidlange6** [comment]
please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/24311	c1f628c
- technical; the implementation is in line with the PR description and the follow up review. This affects only code that was used only if recompiled in a debug mode or when uncommented.
- jenkins tests pass

**Comment by kpedro88** [comment]
+1

**Pull request merged**